firestore/cloud-async-client/snippets.py

```diff
@@ -0,0 +1,774 @@
+    # [START add_custom_class_generated_id_async]
+    city = City(name="Tokyo", state=None, country="Japan")
+    await db.collection("cities").add(city.to_dict())
+    # [END add_custom_class_generated_id_async]
+
+
+async def add_new_doc():
+    db = firestore.AsyncClient()
+    # [START add_new_doc_async]
+    new_city_ref = db.collection("cities").document()
+
+    # later...
+    await new_city_ref.set(
+        {
+            # ...
+        }
+    )
+    # [END add_new_doc_async]
+
+
+async def get_check_exists():
+    db = firestore.AsyncClient()
+    # [START get_check_exists_async]
+    doc_ref = db.collection("cities").document("SF")
+
+    doc = await doc_ref.get()
+    if doc.exists:
+        print(f"Document data: {doc.to_dict()}")
+    else:
+        print("No such document!")
+    # [END get_check_exists_async]
+
+
+async def get_custom_class():
+    db = firestore.AsyncClient()
+    # [START get_custom_class_async]
+    doc_ref = db.collection("cities").document("BJ")
+
+    doc = await doc_ref.get()
+    city = City.from_dict(doc.to_dict())
+    print(city)
+    # [END get_custom_class_async]
+
+
+async def get_simple_query():
+    db = firestore.AsyncClient()
+    # [START get_simple_query_async]
+    # Note: Use of CollectionRef stream() is prefered to get()
+    docs = db.collection("cities").where("capital", "==", True).stream()
+
+    async for doc in docs:
+        print(f"{doc.id} => {doc.to_dict()}")
+    # [END get_simple_query_async]
+
+
+async def array_contains_filter():
+    db = firestore.AsyncClient()
+    # [START fs_array_contains_filter_async]
+    cities_ref = db.collection("cities")
+
+    query = cities_ref.where("regions", "array_contains", "west_coast")
+    # [END fs_array_contains_filter_async]
+    docs = query.stream()
+    async for doc in docs:
+        print(f"{doc.id} => {doc.to_dict()}")
+
+
+async def get_full_collection():
+    db = firestore.AsyncClient()
+    # [START get_full_collection_async]
+    docs = db.collection("cities").stream()
+
+    async for doc in docs:
+        print(f"{doc.id} => {doc.to_dict()}")
+    # [END get_full_collection_async]
+
+
+async def structure_doc_ref():
+    db = firestore.AsyncClient()
+    # [START structure_doc_ref_async]
+    a_lovelace_ref = db.collection("users").document("alovelace")
+    # [END structure_doc_ref_async]
+    print(a_lovelace_ref)
+
+
+async def structure_collection_ref():
+    db = firestore.AsyncClient()
+    # [START structure_collection_ref_async]
+    users_ref = db.collection("users")
+    # [END structure_collection_ref_async]
+    print(users_ref)
+
+
+async def structure_doc_ref_alternate():
+    db = firestore.AsyncClient()
+    # [START structure_doc_ref_alternate_async]
+    a_lovelace_ref = db.document("users/alovelace")
+    # [END structure_doc_ref_alternate_async]
+
+    return a_lovelace_ref
+
+
+async def structure_subcollection_ref():
+    db = firestore.AsyncClient()
+    # [START structure_subcollection_ref_async]
```

Comment:
this sample seems to be identical to sync? Do you have an idea of how common this is?

Reply:
Not terribly common. A couple snippets just create documents/collections locally, of which there are no async calls. I would guess less than 10, maybe around 5. Should we remove these? I felt as though leaving them to illustrate how local creation is the same between sync / async was useful.